### ISSUE TITLE
func-style ignores function expressions not starting with var keyword

### ISSUE THREAD

**Issue by kjholla** [eslint/eslint#7941]
The func-style rule has been set to enforce use of function declarations.

```js
var eslintConfig = {
	'root': true,
	'env': {
		'browser': true
	},
	'extends': 'eslint:recommended',
	'rules': {
		/** Coding Style Standards **/
		'func-style': ['error', 'declaration', { 'allowArrowFunctions': true }]
	}
}

module.exports = eslintConfig;

```

I would expect that with this configuration, code such as the one below would lead to an error. But it doesn't.

```js
var fnVariable;

fnVariable = function() {
      return true;
}

```

**Comment by eslintbot** [comment]
Hi @kjholla, thanks for the issue. It looks like there's not enough information for us to know how to help you.

If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

Requesting a rule change? Please see [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by platinumazure** [comment]
Can't reproduce on the ESLint demo. My guess is you have an old version of ESLint where this was a bug, but upgrading to latest should fix this for you.

In future, please include your ESLint version.

**Comment by kjholla** [comment]
Hello,

Thank you for the response.

My setup is as below:

Node version: 6.2.2
NPM version: 3.9.5
ESLint version: 3.13.1

ESLint Config File (.eslintrc.js):
```
var eslintConfig = {
	'root': true,
	'env': {
		'browser': true
	},
	'extends': 'eslint:recommended',
	'rules': {
		/** Coding Style Standards **/
		'func-style': ['error', 'declaration', { 'allowArrowFunctions': true }]
	}
}

module.exports = eslintConfig;
```


My Source Code:
```
var fnVariable

fnVariable();

fnVariable = function() {
	var a, b;
	a = 2;
	b = 2;
	return a + b;
}
```

What I would have expected: An error as below
`  5:1  error  Expected a function declaration  func-style
`

What I get: No error at all.

**Comment by ljharb** [comment]
In the latter case, if you run it, it will throw a `TypeError` because `fnVariable` is undefined - I'd expect eslint to decide that's intentional, and refuse to warn. If you remove the `fnVariable();` line (ie, any reference to the variable between declaration and initialization), do you get a `func-style` error?

**Comment by kjholla** [comment]
Apologies for the mistake in my example code.

Let us consider something as below.
```
var fnVariable
var doSomethingWithfnVariable;

fnVariable = getSomeValue();
doSomethingWithfnVariable = fnVariable;

fnVariable = function() {
	var a, b;
	a = 2;
	b = 2;
	return a + b;
}

fnVariable();

function getSomeValue() {
	return 2;
}

module.exports = {
	'value': doSomethingWithfnVariable,
	'fn': fnVariable
}
```

And, an ESLint config of:
```
var eslintConfig = {
	'root': true,
	'env': {
		'browser': true
	},
	'globals': {
		'module': true
	},
	'extends': 'eslint:recommended',
	'rules': {
		/** Coding Style Standards **/
		'func-style': ['error', 'declaration', { 'allowArrowFunctions': true }]
	}
}

module.exports = eslintConfig;
```
I would expect ESLint `func-style` rule to warn the developer about the use of a Function Expression being assigned to the variable `fnVariable`. But no such error is seen.

**Comment by ljharb** [comment]
I would not expect an error there - it's impossible to *always* use declarations since sometimes you require an expression. Since your code relies on ordering - ie, relies on not hoisting - using a declaration there isn't possible, since a declaration could break your code.

**Comment by kjholla** [comment]
I would request you to not go strictly with my example code. As a user, I want to be able to enforce strict use of function declarations everywhere. Can this be added as a config option?

**Comment by ljharb** [comment]
That statement would disallow `array.map(function () { })` for example - it's not at all practical to forbid any function expressions.

**Comment by kjholla** [comment]
I see your point, and I agree that disallowing all function expressions may not be practical. However, how about an option to disallow function expressions that are being assigned to variables?

**Comment by ljharb** [comment]
That's what the rule already does, except in cases where it *must* be an expression - such as your example code. I think this is the sort of thing that needs to be caught in code review - if the examples weren't reusing variables (by reassigning to them) then the rule would absolutely be catching them.

**Comment by platinumazure** [comment]
Okay, let's refocus the discussion a bit. Here are my thoughts:

* I can see a use case for returning a lint message when a FunctionExpression is on the right side of an AssignmentExpression, to parallel what happens in a VariableDeclaration. I think that's reasonable and we could treat the current behavior as a bug.
* In cases where literally no FunctionExpressions should be used, ever, we have a different rule for that: `no-restricted-syntax`, which could be enabled on the "FunctionExpression" node type. That will forbid all function expressions entirely (and will allow arrow functions, since those are of type "ArrowFunctionExpression").

@kjholla Could you please clarify exactly what you're looking for here?

**Comment by kjholla** [comment]
@platinumazure You have articulated my concern better than I did myself -- yes, I am concerned about Function Expressions being allowed after Assignment Expressions by the `func-style` rule. I would have expected that setting up a rule `'func-style': ['error', 'declaration', { 'allowArrowFunctions': true }]` would disallow assignment of a function expression to a variable, irrespective of whether or not that variable is being declared in the same statement.

Thanks for that pointer about the `no-restricted-syntax` rule with the `FunctionExpression` option. I will check that one too, though I suspect that it is an all-or-nothing option, which as @ljharb has already pointed out, could be impractical.

**Comment by platinumazure** [comment]
@kjholla Right, I only brought out `no-restricted-syntax` because I wasn't sure just how committed you were to a fight against FunctionExpressions. :smile:

@eslint/eslint-team So the basic issue here is, should `func-style` catch AssignmentExpressions with FunctionExpressions on the right side. I think they should and that the rule as it is is buggy. Anyone have any other thoughts on what we should do here?

**Comment by not-an-aardvark** [comment]
I would say the current behavior is correct -- sometimes it's necessary to reassign a variable to a function without hoisting it.

```js
let foo = 5;

assert.equal(foo + 5, 10);

foo = function() {};
```

Granted, there are better ways to write this code (e.g. using immutable bindings), but if someone does need a mutable reference to a function, using an assignment with a function expression is probably the best way to do it.

**Comment by kjholla** [comment]
That code could also be written by declaring the function separately and then assigning it to the variable foo as and when required.

```
let foo = 5

assert.equal(foo + 5, 10);

foo = _myFunction;

/**
 * JSDoc what the function is meant to do.
 * 
function _myFunction() {
         // Do what it is meant to do...
}

```

**Comment by not-an-aardvark** [comment]
That doesn't work if `_myFunction` creates a closure:

```js
let foo = 5;

assert.equal(foo + 5, 10);

if (bar) {
  let baz = 6;

  foo = function() {
    return baz;
  };
}

foo();
```

**Comment by kjholla** [comment]
OK. I see your point. However, I really would like to understand the scenario where we would have the use for such a closure.

The scenarios I have faced till now have been closures enclosed by a function in itself.

```
let foo = 5;

assert.equal(foo + 5, 10);

if (bar) foo = createClosure();

foo();

function createClosure() {
    let baz = 6;

    return fn;

    function fn() {
        return baz;
    }
}
```

**Comment by not-an-aardvark** [comment]
To answer this question, I tried making the change and running the modified rule on the ESLint codebase. The ESLint codebase violates the modified rule 5 times: 3 times in [`key-spacing`](https://github.com/eslint/eslint/blob/a664e8aefb732b20be68d443eab34700a4436a96/lib/rules/key-spacing.js), and 2 times in [`semi`](https://github.com/eslint/eslint/blob/a664e8aefb732b20be68d443eab34700a4436a96/lib/rules/semi.js) (search for "fix =") to jump to the violations).

**Comment by kjholla** [comment]
Fair point there. And I dare say that I have just learnt some new use for closures. Something I never used before.

However, I also believe that a configuration option to disallow assignment of function expressions would be a valuable addition to the current functionality.

**Comment by p0wdrdotcom** [comment]
Hi those of us that think arrow functions in general are evil can we extend this rule or another to lint all arrow functions out completely?

**Comment by ljharb** [comment]
@p0wdrdotcom i'm going to go on a limb and assume that eslint maintainers would want a bit more of a structured and reasoned argument that a core language feature is "evil"

**Comment by platinumazure** [comment]
@p0wdrdotcom You can use the `no-restricted-syntax` rule to forbid "ArrowFunctionExpression" nodes, that will get you where you need to go.

**Comment by p0wdrdotcom** [comment]
@ljharb I apologise, I should have kept my opinions out of the request. I suppose it would be nice if eslint could lint for any none essential feature of the language to let the developer choose which features they would like to leverage.  I'd like to think that, unlike myself, eslint is an opinion-less tool to be directed by the developer.

Personally, among other things, I think that the implicit return of some arrow functions is ambiguous from a readability perspective and I think I'd rather not use them at all for that fact alone. There are other reasons.  For example the fact that `=>` looks like `>=`  from a far and will sometime creep into code

I would hate to see accidentally get in.
`if (s => 2){
    /* Do something here (always as the function is coerced thru-thy) */
}`

Arrow functions seem to provide less typing and a lexical `this` whilst adding large amounts of ambiguity.  I don't use the lexical this capability as I'm referring to `this` so much less often these days.

**Comment by p0wdrdotcom** [comment]
@platinumazure Thank you. That's exactly what I need.  I couldn't find that exactly by searching.

**Comment by alberto** [comment]
Thanks for your interest in improving eslint. Unfortunately, it looks like this issue didn't get enough support from the team and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend to never do it, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
